File: framework/Castlevania20/Castlevania20/Quadtree.cpp
#include "Quadtree.h"
Quadtree::Quadtree() {}
Quadtree::~Quadtree() {}
void Quadtree::load(char* file) {
	ifstream myfile(file);
	while (!myfile.eof())
	{
		QuadNode* tamp = new QuadNode();
		myfile >> tamp->id;
		myfile >> tamp->bound->X;
		myfile >> tamp->bound->Y;
		myfile >> tamp->bound->Width;
		myfile >> tamp->bound->Height;
		int count = 0;
		myfile >> count;
		for (int i = 0; i < count; i++) {
			int id;
			myfile >> id;
			tamp->listIdObject.push_back(id);
		}
		//list.push_back(tamp);
		if (count > 0) {
			listNode.push_back(tamp);
		}
	}
	for each (QuadNode * b in listNode) {
		if (b->id == 1) {
			_root = b;
			continue;
		}
		int rootId = b->id / 10;
		for each(QuadNode* tamp in listNode) {
			if (tamp->id == rootId) {
				switch (b->id % 10)
				{
				case 1:
					tamp->_nodeTL = b;
					break;
				case 2:
					tamp->_nodeTR = b;
					break;
				case 3:
					tamp->_nodeBL = b;
					break;
				case 4:
					tamp->_nodeBR = b;
					break;
				}
				break;
			}
		}
	}
}


BaseObject* Quadtree::getObject(int type, int x, int y, int width, int height)
{
	BaseObject* tamp = 0;
	switch (type) {

	default: tamp = new BaseObject(1, x, y, width, height); break;
	}
	return tamp;
}


vector<int> Quadtree::Retrieve(CRectangle *rec) {
	vector<int> result;
	for (int i = 0; i < listNode.size(); i++) {
		//if rec colistion with listNode
		//->add item
		if (check(rec, listNode[i]->bound)) {
			for (int j = 0; j < listNode[i]->listIdObject.size(); j++) {
				result.push_back(listNode[i]->listIdObject[j]);
			}
		}
	}
	return result;

}

bool Quadtree::check(CRectangle *rec1, CRectangle *rec2) {
	if ((rec1->X + rec1->Width > rec2->X) && (rec1->X < rec2->X + rec2->Width) ||
		(rec1->Y + rec1->Height > rec2->Y) && (rec1->Y < rec2->Y + rec2->Height)) {
		return true;
	} 
	return false;
}

File: MapEditor/MapEditor/QuadTree.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Drawing;
namespace MapEditor
{
    class QuadTree
    {
        public QuadNode _root;

        public QuadTree(Rectangle bound)
        {
            this._root = new QuadNode(1,0, bound);
        }

        public void buildQuadTree(List<MyObject> listObject)
        {
            this._root.buildQuadNode(listObject);
        }

        public void export(QuadNode node, List<System.Text.StringBuilder> result)
        {
            if (node != null)
            {


                if (node._id == 12)
                {
                    int a = 0;
                }
                //List<string> _Result = new List<string>();
                System.Text.StringBuilder item = new System.Text.StringBuilder("");
                item.Append(String.Format("{0}\t{1}\t{2}\t{3}\t{4}",
                                        node._id.ToString().Trim(),
                                        node._bounds.X.ToString().Trim(),
                                        node._bounds.Y.ToString().Trim(),
                                        node._bounds.Width.ToString().Trim(),
                                        node._bounds.Height.ToString().Trim()));
                item.Append("\t" + node._listObject.Count);

                if (node._nodeBL != null)
                {
                    export(node._nodeBL, result);
                    export(node._nodeBR, result);
                    export(node._nodeTL, result);
                    export(node._nodeTR, result);
                }
                else
                {
                    for (int i = 0; i < node._listObject.Count; ++i)
                    {
                        item.Append("\t" + node._listObject[i].id.ToString().Trim());
                    }
                }
                result.Add(item);
            }
        }
    }
}
